analyzers/src/SonarAnalyzer.Common/Rules/IndexOfCheckAgainstZeroBase.cs

```diff
@@ -0,0 +1,92 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class IndexOfCheckAgainstZeroBase<TSyntaxKind, TExpressionSyntax, TBinaryExpressionSyntax> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TExpressionSyntax : SyntaxNode
+        where TBinaryExpressionSyntax : SyntaxNode
+    {
+        internal static readonly ImmutableArray<KnownType> CheckedTypes =
+            ImmutableArray.Create(
+                KnownType.System_Array,
+                KnownType.System_Collections_Generic_IList_T,
+                KnownType.System_String,
+                KnownType.System_Collections_IList
+            );
+
+        protected const string DiagnosticId = "S2692";
+        private const string MessageFormat = "0 is a valid index, but this check ignores it.";
+
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+        protected abstract TSyntaxKind LessThanExpression { get; }
+        protected abstract TSyntaxKind GreaterThanExpression { get; }
+
+        protected abstract bool TryGetConstantIntValue(TExpressionSyntax expression, out int constValue);
```

Comment:
Because there's more than one usage for it, let's 
* remove this 
* remove `GeneratedCodeRecognizer`
* add ILanguageFacade
* add `ExpressionNumericConverter` to ILanguageFacade and implement it as a `Lazy<ExpressionNumericConverter>`
That should simplify all the scaffolding a lot.